keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+tolerations, effectively disabling the effects. A third potential solution
+involving OCI hooks was then investigated. OCI hooks are separate executables
+that an OCI-compatible container runtime invokes that can modify the behaviour
+of the containers in a sandbox. While this solution would have solved the
+API-layer problem, it introduced other issues, such as the inefficiency of
+downloading the container image to the Node before the OCI hook was run. Since
+pods can be scheduled directly on the node using nodename parameter in pod spec,
+its required to add such validation in Kubelet itself.
+
+The final solution the EKS team settled on involved changing kubelet itself to
+support additional node-local Pod admission checks. This KEP outlines the EKS
+team’s approach and proposes upstreaming these changes to kubelet in order to
+allow extensible node-local last-minute validation checks. This functionality
+will enable cloud providers who support nodeless/serverless worker nodes to
+restrict Pods based on fields other than those already being validated by
+kubelet.
+
+Also, Kubernetes scheduler is extensible but isn’t omniscient about everything
+on the node hence there can be cases where a scheduling decision cannot be
+fulfilled by the node based on the current node conditions such as node-level
+resource allocation. The in-tree PodAdmitHandlers mentioned below are good
+examples for this.
+
+* Eviction manager rejects pods if the node is under high pressure (e.g., PIDs)
+* Topology manager rejects pods if the NUMA affinity cannot be satisfied
+
+There can be additional resource allocation conditions and policies handled
+externally to the kubelet, e.g.,
+
+* Alternative resource allocator implemented by the CRI which needs to reject
+  pods in conditions similar to the built-in resource managers.
+
+These can also benefit from a pluggable admission handler to reject pods before
+they are admitted by the kubelet. This is hard to be replicated by exposing more
+information to the scheduler which may be too costly and subject to race
+conditions.
+
+### Goals
+
+- Allow deployers of fully managed worker nodes to have control over Pods
+  running on those nodes.
+
+- Enable node-local Pod admission checks without requiring changes to kubelet.
+
+### Non-Goals
+
+- Move existing validations to “out of tree” plugins.
+
+- Change existing API-layer validation solutions such as Pod Security Policies
+  and validating admission webhooks.
+
+## Proposal
+
+The approach taken is similar to the container networking interface (CNI) plugin
+architecture. With CNI, kubelet invokes one or more CNI plugin binaries on the
+host to set up a Pod’s networking. kubelet discovers available CNI plugins
+by [examining](https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/kubelet/dockershim/docker_service.go#L242)
+a well-known directory (`/etc/cni/net.d`) for configuration files
```

Comment:
We should define the well defined directory path for the admission plugins as part of this proposal.

Reply:
1.  Ack. We will definitely add and define one. The next iteration may not have it though.

2. A daemonset won't be applicable in this case for these reasons:
A) A cluster administrator can change the properties of the daemonset.
B) By the time a daemonset will run this plugin, it will have run its (daemonset's) container. As a result of this, the daemonset maybe able to get privileged access for Fargate which is what the Plugin intends to prevent.

Note: The intention right now is to  use a unix connection and will work in the same way as node-register plugins.